### PR TITLE
Make data minimization more consistent and performant

### LINKED ISSUE #39: Minimization - Inconsistent NCP calculation
NCP score is calculated on overall generalizations (combined from all cells), this includes when looking for a feature to remove from generalization (also taken from overall generalizations). 
Accuracy is computed on the representative values taken from cells. There can be a mismatch (i.e., generalization applied at cell level but not at overall level) which can lead to not achieving the target accuracy. Need separate NCP calculation for this purpose.

### PR DESCRIPTION
**Pull request by abigailgold**
NCP and accuracy can now be calculated either on global generalization structure or directly on transformed data, according to the required use case scenario and the two are now applied consistently.

Performance of transform method was improved (in case fit was previously called, uses internal decision tree to map records to their appropriate cells).

Fixes #39 and #37.

### CODE REVIEW

File: apt/minimization/minimizer.py [hunk 1]
```diff
@@ -140,24 +160,26 @@ def generalizations(self):
...
     def ncp(self):
         """
-        Return the NCP score of the generalizations.
+        Return the last calculated NCP score. NCP score is calculated upon calling `fit` (on the training data),
+        `transform' (on the test data) or when explicitly calling `calculate_ncp` and providing it a dataset.
 
-        :return: ncp score as float.
+        :return: NCP score as float.
         """
         return self._ncp
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:176]: It's worth having an NCP class, which would return the most appropriate NCP according to the process that was done. Then you won't need the "generalize_using_transform" field to be filled by the user, but only pass it as a parameter to the calculate ncp method, where you know whether transform is being used or not.

File: apt/minimization/minimizer.py [hunk 2]
```diff
@@ -363,27 +376,90 @@ def fit(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, y: Optional[DATA_PANDA
...
+            X_test_dataset = ArrayDataset(X_test, features_names=self._features)
+            self.calculate_ncp(X_test_dataset)
 
         # Return the transformer
         return self
 
     def transform(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, features_names: Optional[list] = None,
                   dataset: Optional[ArrayDataset] = None):
-        """ Transforms data records to representative points.
+        """ Transforms data records to representative points. Updates stored ncp value to the one computed on the
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:387]: Again, according to the NCP-class suggestion, better something like "Adds a new NCP value" to the NCP score.

File: apt/minimization/minimizer.py [hunk 3]
```diff
@@ -363,27 +376,90 @@ def fit(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, y: Optional[DATA_PANDA
...
+        if self._feature_data is None:
+            self._feature_data = self._get_feature_data(samples_pd)
+        total_samples = samples_pd.shape[0]
+
+        if self.generalize_using_transform:
+            generalizations = self._calculate_cell_generalizations()
+            # count how many records are mapped to each cell
+            counted = np.zeros(samples_pd.shape[0])  # to mark records we already counted
+            ncp = 0
+            for i in range(len(self.cells)):
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:441]: Why not iterate over the cells directly? I don't see the index i being used anywhere.

File: apt/minimization/minimizer.py [hunk 4]
```diff
@@ -363,27 +376,90 @@ def fit(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, y: Optional[DATA_PANDA
...
+                cell = self.cells[i]
+                count = self._get_record_count_for_cell(samples_pd, cell, counted)
+                range_counts = {}
+                category_counts = {}
+                for feature in cell['ranges']:
+                    range_counts[feature] = [count]
+                for feature in cell['categories']:
+                    category_counts[feature] = [count]
+                ncp += self._calc_ncp_for_generalization(generalizations[cell['id']], range_counts, category_counts,
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:450]: So, all the cells of the generatlization add up to a single NCP?

File: apt/minimization/minimizer.py [hunk 5]
```diff
@@ -409,40 +486,114 @@ def transform(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, features_names:
...
-    def _cell_contains(self, cell, x, i, mapped):
+    def _get_record_count_for_cell(self, X, cell, mapped):
+        count = 0
+        for index, row in X.iterrows():
+            if not mapped.item(index) and self._cell_contains(cell, row, index, mapped):
+                count += 1
+        return count
+
+    def _cell_contains(self, cell, x, index, mapped):
         for f in self._features:
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:583]: Why not use enumerate() here too to get both the feature and the index together?

File: apt/minimization/minimizer.py [hunk 6]
```diff
@@ -409,40 +486,114 @@ def transform(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, features_names:
...
+                                                     self._feature_data[feature],
+                                                     total_count)
+            total_ncp = total_ncp + feature_ncp
+            total_features += 1
+        if total_features == 0:
+            return 0
+        return total_ncp / total_features
+
+    @staticmethod
+    def _calc_ncp_categorical(categories, category_count, feature_data, total):
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:534]: Maybe feature_category_count**s** and **feature_**category_sizes for consistency?

File: apt/minimization/minimizer.py [hunk 7]
```diff
@@ -409,40 +486,114 @@ def transform(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, features_names:
...
+        category_sizes = [len(g) if len(g) > 1 else 0 for g in categories]
+        normalized_category_sizes = [s * n / total for s, n in zip(category_sizes, category_count)]
+        average_group_size = sum(normalized_category_sizes) / len(normalized_category_sizes)
+        return average_group_size / feature_data['range']  # number of values in category
+
+    @staticmethod
+    def _calc_ncp_numeric(feature_range, range_count, feature_data, total):
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:541]: again, consistency missing between range_count and range_sizes

File: apt/minimization/minimizer.py [hunk 8]
```diff
@@ -813,62 +964,52 @@ def _get_feature_to_remove(self, original_data, prepared_data, nodes, labels, fe
...
                     range_counts[r].append(len(counter))
                     last_value = value
-                counter = [item for item in samples_df[r] if int(item) <= last_value]
+                counter = [item for item in samples[r] if int(item) > last_value]
                 range_counts[r].append(len(counter))
         return range_counts
 
-    def _find_categories_count(self, samples, categories):
+    @staticmethod
+    def _find_categories_counts(samples, categories):
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:1005]: _find_category_counts

**Review by andersonm-ibm — CHANGES_REQUESTED** [review]

File: tests/test_minimizer.py [hunk 9]
```diff
@@ -101,13 +157,194 @@ def test_minimizer_fit():
...
+                   [22, 169]])
+
+    base_est = DecisionTreeClassifier(random_state=0, min_samples_split=2,
+                                      min_samples_leaf=1)
+    model = SklearnClassifier(base_est, ModelOutputType.CLASSIFIER_PROBABILITIES)
+    model.fit(ArrayDataset(X, y))
+    ad = ArrayDataset(X)
+    ad1 = ArrayDataset(X1, features_names=features)
+    predictions = model.predict(ad)
+    if predictions.shape[1] > 1:
```
[the remark this answers is not in the document]
Comment (abigailgold) [reply tests/test_minimizer.py:301]: Extracted a lot of repetitive test code into methods.

File: apt/minimization/minimizer.py (hunk 1 above)
Comment (abigailgold) [reply apt/minimization/minimizer.py]: Added NCPScores class instead of single score.

File: apt/minimization/minimizer.py (hunk 4 above)
Comment (abigailgold) [reply apt/minimization/minimizer.py:470]: Yes, NCP as always global for a complete dataset.

File: apt/minimization/minimizer.py [hunk 10]
```diff
@@ -316,74 +342,143 @@ def fit(self, X: Optional[DATA_PANDAS_NUMPY_TYPE] = None, y: Optional[DATA_PANDA
...
         :type dataset: `ArrayDataset`, optional
         :return: Array containing the representative values to which each record in ``X`` is mapped, as numpy array or
                  pandas DataFrame (depending on the type of ``X``), shape (n_samples, n_features)
         """
+        transformed = self._inner_transform(X, features_names, dataset)
+        transformed_dataset = ArrayDataset(transformed, features_names=self._features)
+        self._ncp_scores.transform_score = self.calculate_ncp(transformed_dataset, True)
+        return transformed
 
+    def calculate_ncp(self, samples: ArrayDataset, generalize_using_transform: bool = True):
```
Comment (andersonm-ibm) [inline apt/minimization/minimizer.py:430]: Does the user still need to fill this `generalize_using_transform` ? 
I thought having NCPScores would enable us here to get the correct NCP score, since we could track the process.

Comment (abigailgold) [reply apt/minimization/minimizer.py:430]: Yes, because some of the stages in the process (like fitting the generalizer) are still affected by this param.

**Review by andersonm-ibm — APPROVED** [review]
Looks good!
